From 577cdba1048ce04816c962264a11efd02f1f5b73 Mon Sep 17 00:00:00 2001
From: Steven Moreland <smoreland@google.com>
Date: Wed, 19 Mar 2025 16:13:37 -0700
Subject: [PATCH] PM: Restrict subclass construction.

Don't construct a random class just because you were asked to.

Bug: 373467684
Test: atest PackageParsingPerfTest
Flag: EXEMPT bugfix
(cherry picked from commit 65c1a90bf4af54f555ded29ec2384072b1c962b8)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:10f7c413fa9176acf1cfcf545b1261e59f128aa9)
Merged-In: I328eca7b8de1eab637e3991adeb2e1d147c30268
Change-Id: I328eca7b8de1eab637e3991adeb2e1d147c30268
---
 core/java/android/content/pm/PackageParser.java | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/core/java/android/content/pm/PackageParser.java b/core/java/android/content/pm/PackageParser.java
index 048289f56a0c..6a7f80c92c97 100644
--- a/core/java/android/content/pm/PackageParser.java
+++ b/core/java/android/content/pm/PackageParser.java
@@ -7763,6 +7763,10 @@ public class PackageParser {
             final ArrayList<T> intentsList;
             try {
                 final Class<T> cls = (Class<T>) Class.forName(componentName);
+                if (!IntentInfo.class.isAssignableFrom(cls)) {
+                    throw new AssertionError("Intent list requires subclass of IntentInfo, not: "
+                            + componentName);
+                }
                 final Constructor<T> cons = cls.getConstructor(Parcel.class);
 
                 intentsList = new ArrayList<>(N);
-- 
2.49.0.1077.gc0e912fd4c-goog

